rust-cross build failure due to missing llvm symbols using `cxx11` references

Doesn't happen on all systems. I've observed this on arch linux, but not on gentoo. Might depend on whether the system gcc has the new cxx11 abi enabled.

Probably due to `uninative.bbclass` from selectively disabling gcc's new cxx11 abi & `cross.bbclass` having a different `*_CXXFLAGS` compared to `native.bbclass`
```
# https://wiki.debian.org/GCC5
# We may see binaries built with gcc5 run or linked into gcc4 environment
# so use the older libstdc++ standard for now until we don't support gcc4
# on the host system.
BUILD_CXXFLAGS_append = " -D_GLIBCXX_USE_CXX11_ABI=0"
```

Full error output
```
| error: linking with `gcc` failed: exit code: 1
| note: "gcc" "-Wl,--enable-new-dtags" "-Wl,--as-needed" "-m64" "-L" "x86_64-linux/stage0/lib/rustlib/x86_64-linux/lib" "x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/bin/rustc.0.o" "-o" "x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/bin/rustc" "-Wl,--gc-sections" "-Wl,-O1" "-nodefaultlibs" "-L" "x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-L" "/home/cody/obj/y/tmp/sysroots/x86_64-linux/usr/lib" "-L" "x86_64-linux/stage0/lib/rustlib/x86_64-linux/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_driver-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_passes-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_borrowck-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_privacy-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_typeck-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_lint-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_resolve-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "syntax_ext-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_plugin-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_metadata-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_trans-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_platform_intrinsics-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_mir-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_data_structures-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_back-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "flate-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "getopts-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "arena-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "fmt_macros-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "graphviz-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_front-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "syntax-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "term-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rustc_llvm-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "rbml-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "serialize-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "log-ca1c970e" "-L" "/home/cody/obj/y/tmp/work/x86_64-linux/rust-cross-arm/1.7.0-r0/build/x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib" "-l" "std-ca1c970e" "-l" "pthread" "-l" "edit" "-l" "dl" "-l" "m" "-l" "stdc++" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-Wl,-rpath,$ORIGIN/../lib" "-Wl,-rpath,/usr/local/lib/rustlib/x86_64-linux/lib" "-l" "compiler-rt" "-L/home/cody/obj/y/tmp/sysroots/x86_64-linux/usr/lib" "-L/home/cody/obj/y/tmp/sysroots/x86_64-linux/lib" "-Wl,-rpath-link,/home/cody/obj/y/tmp/sysroots/x86_64-linux/usr/lib" "-Wl,-rpath-link,/home/cody/obj/y/tmp/sysroots/x86_64-linux/lib" "-Wl,-rpath,/home/cody/obj/y/tmp/sysroots/x86_64-linux/usr/lib" "-Wl,-rpath,/home/cody/obj/y/tmp/sysroots/x86_64-linux/lib" "-Wl,-O1" "-Wl,-rpath=../../lib"
| note: x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib/librustc_llvm-ca1c970e.so: undefined reference to `llvm::TargetRegistry::lookupTarget(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'
| x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib/librustc_llvm-ca1c970e.so: undefined reference to `llvm::RTDyldMemoryManager::getSymbolAddressInProcess(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
| x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib/librustc_llvm-ca1c970e.so: undefined reference to `llvm::sys::DynamicLibrary::getPermanentLibrary(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)'
| x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib/librustc_llvm-ca1c970e.so: undefined reference to `llvm::createPrintModulePass(llvm::raw_ostream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
| x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib/librustc_llvm-ca1c970e.so: undefined reference to `llvm::report_fatal_error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
| x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib/librustc_llvm-ca1c970e.so: undefined reference to `llvm::SectionMemoryManager::finalizeMemory(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)'
| x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib/librustc_llvm-ca1c970e.so: undefined reference to `llvm::Triple::normalize[abi:cxx11](llvm::StringRef)'
| x86_64-linux/stage0/lib/arm-poky-linux-gnueabi/rustlib/x86_64-linux/lib/librustc_llvm-ca1c970e.so: undefined reference to `llvm::RTDyldMemoryManager::getPointerToNamedFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
```

--- a/recipes-devtools/rust/rust.inc
+++ b/recipes-devtools/rust/rust.inc
@@ -359,6 +359,13 @@ def rust_gen_mk_cfg(d, thing):
     if r:
         raise Exception
     o.write("OBJCOPY_{} := {}objcopy\n".format(sys, prefix))
+    # Note: this isn't how this variable is supposed to be used, but for
+    # non-msvc platforms nothing else touches it.
+    # These are the only extra flags passed to the rustllvm (c++ code) build.
+    # These are only used for host (even though we emit them for all targets)
+    # Without this, there are link failures due to GLIBC_CXX11_ABI issues in
+    # certain setups.
+    o.write("EXTRA_RUSTLLVM_CXXFLAGS_{} := {}\n".format(sys, cxxflags_for(d, thing)))
     o.close()
     i.close()
 

--- a/recipes-devtools/rust/rust.inc
+++ b/recipes-devtools/rust/rust.inc
@@ -247,8 +247,10 @@ def rust_gen_target(d, thing, wd):
     prefix = prefix_for(d, thing)
 
     features = ""
+    cpu = "generic"
     if thing is "TARGET":
         features = d.getVar('TARGET_LLVM_FEATURES', True) or ""
+        cpu = d.getVar('TARGET_LLVM_CPU', True)
     features = features or d.getVarFlag('FEATURES', arch, True) or ""
 
     # build tspec
@@ -264,7 +266,7 @@ def rust_gen_target(d, thing, wd):
     tspec['linker'] = "{}{}gcc".format(d.getVar('CCACHE', True), prefix)
     tspec['objcopy'] = "{}objcopy".format(prefix)
     tspec['ar'] = "{}ar".format(prefix)
-    tspec['cpu'] = d.getVar('TARGET_LLVM_CPU', True)
+    tspec['cpu'] = cpu
     if features is not "":
         tspec['features'] = features
     tspec['dynamic-linking'] = True

--- a/recipes-devtools/rust/rust.inc
+++ b/recipes-devtools/rust/rust.inc
@@ -238,12 +238,9 @@ def ldflags_for(d, thing, arch):
 
 TARGET_LLVM_CPU="${@llvm_cpu(d)}"
 TARGET_LLVM_FEATURES = "${@llvm_features_from_tune(d)} ${@llvm_features_from_cc_arch(d)}"
-TARGET_LLVM_CPU_class-cross="${@llvm_cpu(d)}"
-TARGET_LLVM_FEATURES_class-cross  = "${@llvm_features_from_tune(d)} ${@llvm_features_from_cc_arch(d)}"
 
 # class-native implies TARGET=HOST, and TUNE_FEATURES only describes the real
 # (original) target.
-TARGET_LLVM_CPU="${@llvm_cpu(d)}"
 TARGET_LLVM_FEATURES_class-native = "${@llvm_features_from_cc_arch(d)}"
 
 def rust_gen_target(d, thing, wd):

--- a/recipes-devtools/rust/rust.inc
+++ b/recipes-devtools/rust/rust.inc
@@ -13,9 +13,6 @@ B = "${WORKDIR}/build"
 
 DEPENDS += "file-native"
 
-# Avoid having the default bitbake.conf disable sub-make parallelization
-EXTRA_OEMAKE = ""
-
 PACKAGECONFIG ??= ""
 
 # Controls whether we use the local rust to build.

--- a/recipes-devtools/rust/rust.inc
+++ b/recipes-devtools/rust/rust.inc
@@ -320,6 +320,15 @@ def rust_gen_mk_cfg(d, thing):
     i = open(d.getVar('S', True) + '/mk/cfg/' + rust_base_sys + '.mk', 'r')
 
     r = subprocess.call(['sed',
+        # CFLAGS, LDFLAGS, CXXFLAGS, CPPFLAGS are used by rust's build for a
+        # wide range of targets (not just HOST). Yocto's settings for them will
+        # be inappropriate, avoid having random targets try to use them, we'll
+        # add as needed.
+        '-e', 's/$(CFLAGS)//',
+        '-e', 's/$(CXXFLAGS)//',
+        '-e', 's/$(CPPFLAGS)//',
+        '-e', 's/$(LDFLAGS)//',
+
         # update all triplets to the new one
         '-e', 's/{}/{}/g'.format(rust_base_sys, sys),
 
@@ -376,14 +385,6 @@ do_configure () {
     cd "${B}"
 
     # FIXME: target_prefix vs prefix, see cross.bbclass
-
-    # CFLAGS, LDFLAGS, CXXFLAGS, CPPFLAGS are used by rust's build for a
-    # wide range of targets (not just HOST). Yocto's settings for them will
-    # be inappropriate, avoid using.
-    unset CFLAGS
-    unset LDFLAGS
-    unset CXXFLAGS
-    unset CPPFLAGS
 
     # FIXME: this path to rustc (via `which rustc`) may not be quite right in the case
     # where we're reinstalling the compiler. May want to try for a real

